File: sdks/kotlin/src/main/kotlin/ly/neptune/astro/payments/PaymentsClient.kt
package ly.neptune.astro.payments

import ly.neptune.astro.HttpEngine
import ly.neptune.astro.models.*

class PaymentsClient internal constructor(private val engine: HttpEngine) {

    suspend fun createSession(request: CreateSessionRequest, idempotencyKey: String? = null): PaymentSession =
        engine.post(
            "/payments/initiate",
            request,
            idempotencyKey?.let { mapOf("Idempotency-Key" to it) } ?: emptyMap()
        )

    suspend fun getSession(sessionId: String): PaymentSession =
        engine.get("/payments/$sessionId")

    suspend fun listSessions(
        status: String? = null,
        cursor: String? = null,
        limit: Int? = null
    ): ListSessionsResponse = engine.get(
        "/payments",
        mapOf("status" to status, "cursor" to cursor, "limit" to limit?.toString())
    )

    suspend fun cancelSession(sessionId: String): PaymentSession =
        engine.post("/payments/$sessionId/cancel")

    suspend fun createRefund(sessionId: String, request: CreateRefundRequest, idempotencyKey: String? = null): Refund =
        engine.post(
            "/payments/$sessionId/refunds",
            request,
            idempotencyKey?.let { mapOf("Idempotency-Key" to it) } ?: emptyMap()
        )

    suspend fun listRefunds(sessionId: String): RefundListResponse =
        engine.get("/payments/$sessionId/refunds")

    suspend fun getRefund(refundId: String): Refund =
        engine.get("/payments/refunds/$refundId")

    suspend fun createMandate(request: CreateMandateRequest, idempotencyKey: String? = null): Mandate =
        engine.post(
            "/recurring/mandates",
            request,
            idempotencyKey?.let { mapOf("Idempotency-Key" to it) } ?: emptyMap()
        )

    suspend fun listMandates(
        status: String? = null,
        page: Int? = null,
        limit: Int? = null
    ): MandateListResponse = engine.get(
        "/recurring/mandates",
        mapOf("status" to status, "page" to page?.toString(), "limit" to limit?.toString())
    )

    suspend fun getMandate(mandateId: String): Mandate =
        engine.get("/recurring/mandates/$mandateId")

    suspend fun cancelMandate(mandateId: String): Mandate =
        engine.delete("/recurring/mandates/$mandateId")

    suspend fun chargeMandate(
        mandateId: String,
        request: ChargeMandateRequest,
        idempotencyKey: String? = null
    ): MandateCharge = engine.post(
        "/recurring/mandates/$mandateId/charge",
        request,
        idempotencyKey?.let { mapOf("Idempotency-Key" to it) } ?: emptyMap()
    )

    suspend fun listMandateCharges(
        mandateId: String,
        page: Int? = null,
        limit: Int? = null
    ): MandateChargeListResponse = engine.get(
        "/recurring/mandates/$mandateId/charges",
        mapOf("page" to page?.toString(), "limit" to limit?.toString())
    )
}

File: sdks/swift/Sources/AstroSDK/Payments/PaymentsClient.swift
import Foundation

public final class PaymentsClient {
    private let http: AstroHTTPClient
    init(http: AstroHTTPClient) { self.http = http }

    public func createSession(_ request: CreateSessionRequest, idempotencyKey: String? = nil) async throws -> PaymentSession {
        let headers = idempotencyKey.map { ["Idempotency-Key": $0] } ?? [:]
        let req = http.request(http.url("/payments/initiate"), method: "POST", body: request, extraHeaders: headers)
        return try await http.perform(req)
    }

    public func getSession(_ sessionId: String) async throws -> PaymentSession {
        let req = http.request(http.url("/payments/\(sessionId)"))
        return try await http.perform(req)
    }

    public func cancelSession(_ sessionId: String) async throws -> PaymentSession {
        let req = http.request(http.url("/payments/\(sessionId)/cancel"), method: "POST")
        return try await http.perform(req)
    }

    public func createRefund(
        sessionId: String,
        request: CreateRefundRequest,
        idempotencyKey: String? = nil
    ) async throws -> Refund {
        let headers = idempotencyKey.map { ["Idempotency-Key": $0] } ?? [:]
        let req = http.request(
            http.url("/payments/\(sessionId)/refunds"),
            method: "POST",
            body: request,
            extraHeaders: headers
        )
        return try await http.perform(req)
    }

    public func listRefunds(sessionId: String) async throws -> RefundListResponse {
        let req = http.request(http.url("/payments/\(sessionId)/refunds"))
        return try await http.perform(req)
    }

    public func getRefund(_ refundId: String) async throws -> Refund {
        let req = http.request(http.url("/payments/refunds/\(refundId)"))
        return try await http.perform(req)
    }

    public func listSessions(status: String? = nil, cursor: String? = nil) async throws -> [PaymentSession] {
        let url = http.url("/payments", query: ["status": status, "cursor": cursor])
        let req = http.request(url)
        let result = try await http.perform(req) as ListSessionsWrapper
        return result.sessions
    }

    public func createMandate(_ request: CreateMandateRequest, idempotencyKey: String? = nil) async throws -> Mandate {
        let headers = idempotencyKey.map { ["Idempotency-Key": $0] } ?? [:]
        let req = http.request(http.url("/recurring/mandates"), method: "POST", body: request, extraHeaders: headers)
        return try await http.perform(req)
    }

    public func listMandates(status: String? = nil, page: Int? = nil, limit: Int? = nil) async throws -> MandateListResponse {
        let url = http.url("/recurring/mandates", query: [
            "status": status,
            "page": page.map(String.init),
            "limit": limit.map(String.init)
        ])
        let req = http.request(url)
        return try await http.perform(req)
    }

    public func getMandate(_ mandateId: String) async throws -> Mandate {
        let req = http.request(http.url("/recurring/mandates/\(mandateId)"))
        return try await http.perform(req)
    }

    public func cancelMandate(_ mandateId: String) async throws -> Mandate {
        let req = http.request(http.url("/recurring/mandates/\(mandateId)"), method: "DELETE")
        return try await http.perform(req)
    }

    public func chargeMandate(
        _ mandateId: String,
        request: ChargeMandateRequest,
        idempotencyKey: String? = nil
    ) async throws -> MandateCharge {
        let headers = idempotencyKey.map { ["Idempotency-Key": $0] } ?? [:]
        let req = http.request(
            http.url("/recurring/mandates/\(mandateId)/charge"),
            method: "POST",
            body: request,
            extraHeaders: headers
        )
        return try await http.perform(req)
    }

    public func listMandateCharges(_ mandateId: String, page: Int? = nil, limit: Int? = nil) async throws -> MandateChargeListResponse {
        let url = http.url("/recurring/mandates/\(mandateId)/charges", query: [
            "page": page.map(String.init),
            "limit": limit.map(String.init)
        ])
        let req = http.request(url)
        return try await http.perform(req)
    }
}

private struct ListSessionsWrapper: Decodable { let sessions: [PaymentSession] }
